File: backend/internal/model/ad_studio.go
package model

import (
	"strings"
	"time"
)

const (
	AdStudioCatalogStudio = "studio"
	AdStudioCatalogTrends = "trends"

	AdStudioCategoryProductShot = "product_shot"
	AdStudioCategoryMotion      = "motion"
	AdStudioCategoryUGC         = "ugc"
	AdStudioCategoryAds         = "ads"
	AdStudioCategoryPosters     = "posters"
	AdStudioCategoryMarketplace = "marketplace"

	AdTrendsCategoryViral      = "viral"
	AdTrendsCategoryMemes      = "memes"
	AdTrendsCategoryChallenges = "challenges"
	AdTrendsCategorySeasonal   = "seasonal"
	AdTrendsCategoryNews       = "news"
	AdTrendsCategoryFormats    = "formats"

	AdStudioMediaImage = "image"
	AdStudioMediaVideo = "video"

	AdStudioModeTextToImage      = "text-to-image"
	AdStudioModeImageToImage     = "image-to-image"
	AdStudioModeCombine          = "combine"
	AdStudioModeTextToVideo      = "text-to-video"
	AdStudioModeImageToVideo     = "image-to-video"
	AdStudioModeReferenceToVideo = "reference-to-video"
)

func NormalizeAdStudioGenerationMode(mode string) string {
	switch strings.TrimSpace(mode) {
	case AdStudioModeTextToImage, AdStudioModeImageToImage, AdStudioModeCombine,
		AdStudioModeTextToVideo, AdStudioModeImageToVideo, AdStudioModeReferenceToVideo:
		return strings.TrimSpace(mode)
	default:
		return ""
	}
}

func AdStudioMediaKindForMode(mode string) string {
	switch NormalizeAdStudioGenerationMode(mode) {
	case AdStudioModeTextToVideo, AdStudioModeImageToVideo, AdStudioModeReferenceToVideo:
		return AdStudioMediaVideo
	default:
		return AdStudioMediaImage
	}
}

func AdStudioModeNeedsProduct(mode string) bool {
	switch NormalizeAdStudioGenerationMode(mode) {
	case AdStudioModeImageToImage, AdStudioModeCombine, AdStudioModeImageToVideo, AdStudioModeReferenceToVideo:
		return true
	default:
		return false
	}
}

func AdStudioModeUsesTemplateInput(mode string) bool {
	switch NormalizeAdStudioGenerationMode(mode) {
	case AdStudioModeCombine, AdStudioModeReferenceToVideo:
		return true
	default:
		return false
	}
}

func NormalizeAdStudioCatalog(catalog string) string {
	switch strings.TrimSpace(catalog) {
	case AdStudioCatalogTrends:
		return AdStudioCatalogTrends
	default:
		return AdStudioCatalogStudio
	}
}

type AdStudioTemplate struct {
	ID                 string
	Title              string
	Description        string
	Catalog            string
	Category           string
	MediaKind          string
	GenerationMode     string
	AspectRatio        string
	Duration           int
	SystemPrompt       string
	PreviewS3Key       string
	PreviewContentType string
	PreviewThumbS3Key  string
	RequiresProduct    bool
	RequiresAvatar     bool
	SortOrder          int
	IsPublished        bool
	CreatedAt          time.Time
	UpdatedAt          time.Time
}

type AdStudioCategoryView struct {
	ID    string `json:"id"`
	Label string `json:"label"`
}

var AdStudioCategories = []AdStudioCategoryView{
	{ID: AdStudioCategoryProductShot, Label: "Съёмка товара"},
	{ID: AdStudioCategoryMotion, Label: "Движение"},
	{ID: AdStudioCategoryUGC, Label: "UGC"},
	{ID: AdStudioCategoryAds, Label: "Реклама"},
	{ID: AdStudioCategoryPosters, Label: "Постеры"},
	{ID: AdStudioCategoryMarketplace, Label: "Маркетплейс"},
}

var AdTrendsCategories = []AdStudioCategoryView{
	{ID: AdTrendsCategoryViral, Label: "Вирусное"},
	{ID: AdTrendsCategoryMemes, Label: "Мемы"},
	{ID: AdTrendsCategoryChallenges, Label: "Челленджи"},
	{ID: AdTrendsCategorySeasonal, Label: "Сезонное"},
	{ID: AdTrendsCategoryNews, Label: "Новости"},
	{ID: AdTrendsCategoryFormats, Label: "Форматы"},
}

func CategoriesForCatalog(catalog string) []AdStudioCategoryView {
	if NormalizeAdStudioCatalog(catalog) == AdStudioCatalogTrends {
		return AdTrendsCategories
	}
	return AdStudioCategories
}

func IsAdStudioCategory(id string) bool {
	return categoryInList(id, AdStudioCategories)
}

func IsAdTrendsCategory(id string) bool {
	return categoryInList(id, AdTrendsCategories)
}

func categoryInList(id string, items []AdStudioCategoryView) bool {
	id = strings.TrimSpace(id)
	for _, item := range items {
		if item.ID == id {
			return true
		}
	}
	return false
}

func AdStudioCategoryLabel(id string) string {
	for _, item := range AdStudioCategories {
		if item.ID == id {
			return item.Label
		}
	}
	for _, item := range AdTrendsCategories {
		if item.ID == id {
			return item.Label
		}
	}
	return id
}

func VisibleAdStudioCategories(hidden []string) []AdStudioCategoryView {
	return VisibleCategoriesForCatalog(AdStudioCatalogStudio, hidden)
}

func VisibleCategoriesForCatalog(catalog string, hidden []string) []AdStudioCategoryView {
	blocked := make(map[string]bool, len(hidden))
	for _, id := range hidden {
		blocked[strings.TrimSpace(id)] = true
	}
	src := CategoriesForCatalog(catalog)
	out := make([]AdStudioCategoryView, 0, len(src))
	for _, item := range src {
		if !blocked[item.ID] {
			out = append(out, item)
		}
	}
	return out
}

type AdStudioTemplatePublicView struct {
	ID               string `json:"id"`
	Title            string `json:"title"`
	Description      string `json:"description"`
	Catalog          string `json:"catalog,omitempty"`
	Category         string `json:"category"`
	MediaKind        string `json:"media_kind"`
	GenerationMode   string `json:"generation_mode"`
	AspectRatio      string `json:"aspect_ratio"`
	Duration         int    `json:"duration"`
	RequiresProduct  bool   `json:"requires_product"`
	RequiresAvatar   bool   `json:"requires_avatar"`
	PreviewKind      string `json:"preview_kind,omitempty"`
	PreviewURL       string `json:"preview_url,omitempty"`
	PreviewSourceURL string `json:"preview_source_url,omitempty"`
	SortOrder        int    `json:"sort_order"`
}

type AdStudioTemplateAdminView struct {
	AdStudioTemplatePublicView
	SystemPrompt string `json:"system_prompt"`
	IsPublished  bool   `json:"is_published"`
	HasPreview   bool   `json:"has_preview"`
	CreatedAt    string `json:"created_at"`
	UpdatedAt    string `json:"updated_at"`
}

type AdStudioTemplateWriteRequest struct {
	Title           string `json:"title"`
	Description     string `json:"description"`
	Catalog         string `json:"catalog"`
	Category        string `json:"category"`
	MediaKind       string `json:"media_kind"`
	GenerationMode  string `json:"generation_mode"`
	AspectRatio     string `json:"aspect_ratio"`
	Duration        int    `json:"duration"`
	SystemPrompt    string `json:"system_prompt"`
	RequiresProduct *bool  `json:"requires_product"`
	RequiresAvatar  *bool  `json:"requires_avatar"`
	SortOrder       *int   `json:"sort_order"`
	IsPublished     *bool  `json:"is_published"`
}

type AdStudioGenerateRequest struct {
	ProductUploadID string `json:"product_upload_id"`
	AvatarUploadID  string `json:"avatar_upload_id,omitempty"`
	Edit            string `json:"edit"`
}

func AdStudioPreviewIsVideo(contentType string) bool {
	return strings.HasPrefix(strings.ToLower(strings.TrimSpace(contentType)), "video/")
}

func (t AdStudioTemplate) PreviewPath() string {
	if t.PreviewS3Key == "" {
		return ""
	}
	return "/ad-studio/templates/" + t.ID + "/preview"
}

func (t AdStudioTemplate) PreviewSourcePath() string {
	if t.PreviewS3Key == "" || !AdStudioPreviewIsVideo(t.PreviewContentType) {
		return ""
	}
	return "/ad-studio/templates/" + t.ID + "/preview/source"
}

func (t AdStudioTemplate) CatalogPreviewPath() string {
	if t.PreviewS3Key == "" {
		return ""
	}
	return "/public/ad-studio/templates/" + t.ID + "/preview"
}

func (t AdStudioTemplate) CatalogPreviewSourcePath() string {
	if t.PreviewS3Key == "" || !AdStudioPreviewIsVideo(t.PreviewContentType) {
		return ""
	}
	return "/public/ad-studio/templates/" + t.ID + "/preview/source"
}

func (t AdStudioTemplate) ToPublicView() AdStudioTemplatePublicView {
	previewKind := AdStudioMediaImage
	if AdStudioPreviewIsVideo(t.PreviewContentType) {
		previewKind = AdStudioMediaVideo
	}
	view := AdStudioTemplatePublicView{
		ID:              t.ID,
		Title:           t.Title,
		Description:     t.Description,
		Catalog:         NormalizeAdStudioCatalog(t.Catalog),
		Category:        t.Category,
		MediaKind:       t.MediaKind,
		GenerationMode:  t.GenerationMode,
		AspectRatio:     t.AspectRatio,
		Duration:        t.Duration,
		RequiresProduct: t.RequiresProduct,
		RequiresAvatar:  t.RequiresAvatar,
		PreviewKind:     previewKind,
		PreviewURL:      t.PreviewPath(),
		SortOrder:       t.SortOrder,
	}
	if previewKind == AdStudioMediaVideo {
		view.PreviewSourceURL = t.PreviewSourcePath()
	}
	return view
}

func (t AdStudioTemplate) ToCatalogView() AdStudioTemplatePublicView {
	view := t.ToPublicView()
	view.PreviewURL = t.CatalogPreviewPath()
	if view.PreviewKind == AdStudioMediaVideo {
		view.PreviewSourceURL = t.CatalogPreviewSourcePath()
	}
	return view
}

func (t AdStudioTemplate) ToAdminView() AdStudioTemplateAdminView {
	return AdStudioTemplateAdminView{
		AdStudioTemplatePublicView: t.ToPublicView(),
		SystemPrompt:               t.SystemPrompt,
		IsPublished:                t.IsPublished,
		HasPreview:                 t.PreviewS3Key != "",
		CreatedAt:                  t.CreatedAt.UTC().Format(time.RFC3339),
		UpdatedAt:                  t.UpdatedAt.UTC().Format(time.RFC3339),
	}
}

File: frontend/src/lib/ad-studio.ts
import { ApiError, apiFetch } from "@/lib/api";
import { postGenerationMultipart } from "@/lib/generation-upload";
import type { GenerationJob } from "@/lib/generation-api";
import type { VideoGenerationJob } from "@/lib/video-generation-api";
import {
  REFERENCE_VIDEO_MAX_BYTES,
  isReferenceVideoDurationValid,
  referenceVideoMaxAllowedSeconds,
} from "@/lib/video-generation-data";

export const AD_STUDIO_CATEGORIES = [
  { id: "product_shot", label: "Съёмка товара" },
  { id: "motion", label: "Движение" },
  { id: "ugc", label: "UGC" },
  { id: "ads", label: "Реклама" },
  { id: "posters", label: "Постеры" },
  { id: "marketplace", label: "Маркетплейс" },
] as const;

export type AdStudioCategoryId = (typeof AD_STUDIO_CATEGORIES)[number]["id"];
export type AdStudioMediaKind = "image" | "video";
export type AdStudioGenerationMode =
  | "text-to-image"
  | "image-to-image"
  | "combine"
  | "text-to-video"
  | "image-to-video"
  | "reference-to-video";

export const AD_STUDIO_GENERATION_MODES: {
  id: AdStudioGenerationMode;
  label: string;
  desc: string;
  mediaKind: AdStudioMediaKind;
}[] = [
  {
    id: "text-to-image",
    label: "Текст → фото",
    desc: "Только описание. Превью шаблона не уходит в модель.",
    mediaKind: "image",
  },
  {
    id: "image-to-image",
    label: "Фото → фото",
    desc: "На входе фото товара. Стиль берётся из промпта шаблона.",
    mediaKind: "image",
  },
  {
    id: "combine",
    label: "Комбинация фото",
    desc: "Превью шаблона + фото товара. Товар встаёт в сцену шаблона.",
    mediaKind: "image",
  },
  {
    id: "text-to-video",
    label: "Текст → видео",
    desc: "Только описание. Превью шаблона не уходит в модель.",
    mediaKind: "video",
  },
  {
    id: "image-to-video",
    label: "Фото → видео",
    desc: "На входе фото товара. Из него собирается ролик.",
    mediaKind: "video",
  },
  {
    id: "reference-to-video",
    label: "Референс → видео",
    desc: "Превью шаблона + фото товара как референсы сцены.",
    mediaKind: "video",
  },
];

export type AdStudioTemplate = {
  id: string;
  title: string;
  description: string;
  category: AdStudioCategoryId;
  media_kind: AdStudioMediaKind;
  generation_mode: AdStudioGenerationMode;
  aspect_ratio: string;
  duration: number;
  requires_product: boolean;
  requires_avatar: boolean;
  preview_kind?: AdStudioMediaKind;
  preview_url?: string;
  preview_source_url?: string;
  sort_order: number;
};

export type AdStudioTemplateAdmin = AdStudioTemplate & {
  system_prompt: string;
  is_published: boolean;
  has_preview: boolean;
  created_at: string;
  updated_at: string;
};

export type AdStudioWritePayload = {
  title: string;
  description: string;
  category: AdStudioCategoryId;
  media_kind: AdStudioMediaKind;
  generation_mode: AdStudioGenerationMode;
  aspect_ratio: string;
  duration: number;
  system_prompt: string;
  requires_product: boolean;
  requires_avatar: boolean;
  sort_order: number;
  is_published: boolean;
};

export function adStudioCategoryLabel(id: string): string {
  return AD_STUDIO_CATEGORIES.find((c) => c.id === id)?.label ?? id;
}

export function defaultAdStudioKind(category: AdStudioCategoryId): AdStudioMediaKind {
  return adStudioMediaKindForMode(defaultAdStudioMode(category));
}

export function defaultAdStudioMode(category: AdStudioCategoryId): AdStudioGenerationMode {
  return category === "motion" || category === "ugc" ? "reference-to-video" : "combine";
}

export function adStudioMediaKindForMode(mode: AdStudioGenerationMode): AdStudioMediaKind {
  return AD_STUDIO_GENERATION_MODES.find((item) => item.id === mode)?.mediaKind ?? "image";
}

export function adStudioModeLabel(mode: string): string {
  return AD_STUDIO_GENERATION_MODES.find((item) => item.id === mode)?.label ?? mode;
}

export function adStudioModeNeedsProduct(mode: AdStudioGenerationMode): boolean {
  return (
    mode === "image-to-image" ||
    mode === "combine" ||
    mode === "image-to-video" ||
    mode === "reference-to-video"
  );
}

export function adStudioModeUsesTemplateInput(mode: AdStudioGenerationMode): boolean {
  return mode === "combine" || mode === "reference-to-video";
}

export function resolveAdStudioMode(item: {
  generation_mode?: string;
  media_kind?: string;
}): AdStudioGenerationMode {
  const mode = AD_STUDIO_GENERATION_MODES.find((m) => m.id === item.generation_mode)?.id;
  if (mode) return mode;
  return item.media_kind === "video" ? "reference-to-video" : "combine";
}

export function defaultAdStudioRatio(
  category: AdStudioCategoryId,
  kind: AdStudioMediaKind,
): string {
  if (kind === "video") return "9:16";
  if (category === "posters" || category === "product_shot") return "4:5";
  return "1:1";
}

export function visibleAdStudioCategories(hidden: string[] | undefined) {
  const blocked = new Set(hidden ?? []);
  return AD_STUDIO_CATEGORIES.filter((item) => !blocked.has(item.id));
}

export function fetchAdStudioTemplates(category?: string) {
  const qs = category ? `?category=${encodeURIComponent(category)}` : "";
  return apiFetch<{ items: AdStudioTemplate[]; hidden_categories?: string[] }>(
    `/ad-studio/templates${qs}`,
  );
}

export function fetchAdminAdStudioCategories() {
  return apiFetch<{ hidden_categories: string[]; shuffle_templates?: boolean }>(
    "/admin/ad-studio/categories",
  );
}

export function updateAdminAdStudioCategories(hidden: string[], shuffleTemplates: boolean) {
  return apiFetch<{ hidden_categories: string[]; shuffle_templates: boolean }>(
    "/admin/ad-studio/categories",
    {
      method: "PUT",
      body: JSON.stringify({
        hidden_categories: hidden,
        shuffle_templates: shuffleTemplates,
      }),
    },
  );
}

export function generateFromAdStudioTemplate(
  id: string,
  body: { product_upload_id?: string; avatar_upload_id?: string; edit?: string },
) {
  return apiFetch<{
    job: GenerationJob | VideoGenerationJob;
    media_kind: AdStudioMediaKind;
  }>(`/ad-studio/templates/${encodeURIComponent(id)}/generate`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function fetchAdminAdStudioTemplates(category?: string) {
  const qs = category ? `?category=${encodeURIComponent(category)}` : "";
  return apiFetch<{ items: AdStudioTemplateAdmin[] }>(
    `/admin/ad-studio/templates${qs}`,
  );
}

export function createAdminAdStudioTemplate(payload: AdStudioWritePayload) {
  return apiFetch<{ item: AdStudioTemplateAdmin }>("/admin/ad-studio/templates", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateAdminAdStudioTemplate(
  id: string,
  payload: AdStudioWritePayload,
) {
  return apiFetch<{ item: AdStudioTemplateAdmin }>(
    `/admin/ad-studio/templates/${encodeURIComponent(id)}`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
    },
  );
}

export function deleteAdminAdStudioTemplate(id: string) {
  return apiFetch<{ ok: boolean }>(
    `/admin/ad-studio/templates/${encodeURIComponent(id)}`,
    { method: "DELETE" },
  );
}

export async function uploadAdminAdStudioPreview(id: string, file: File) {
  const form = new FormData();
  form.append("file", file);
  try {
    return await postGenerationMultipart<{ item: AdStudioTemplateAdmin }>(
      `/admin/ad-studio/templates/${encodeURIComponent(id)}/preview`,
      form,
    );
  } catch (err) {
    if (err instanceof ApiError) throw err;
    throw new ApiError(0, "Не удалось загрузить превью");
  }
}

export function adminAdStudioPreviewUrl(item: AdStudioTemplateAdmin): string {
  if (!item.has_preview) return "";
  return `/admin/ad-studio/templates/${encodeURIComponent(item.id)}/preview?t=${encodeURIComponent(item.updated_at)}`;
}

export function adminAdStudioPreviewSourceUrl(item: AdStudioTemplateAdmin): string {
  if (!item.has_preview || item.preview_kind !== "video") return "";
  return `/admin/ad-studio/templates/${encodeURIComponent(item.id)}/preview/source?t=${encodeURIComponent(item.updated_at)}`;
}

export function validateAdStudioPreviewFile(
  file: File,
  mediaKind: AdStudioMediaKind,
): string | null {
  if (file.type.startsWith("video/")) {
    if (mediaKind !== "video") {
      return "Видео-превью доступно только для видео-шаблонов";
    }
    if (file.size > REFERENCE_VIDEO_MAX_BYTES) {
      return "Видео должно быть не больше 50 МБ";
    }
    return null;
  }
  if (file.type.startsWith("image/")) {
    if (file.size > 15 * 1024 * 1024) {
      return "Изображение должно быть не больше 15 МБ";
    }
    return null;
  }
  return "Поддерживаются фото (JPEG, PNG, WebP) или видео (MP4, MOV, WebM)";
}

export async function probeAdStudioVideoDuration(file: File): Promise<number | null> {
  return new Promise((resolve) => {
    const url = URL.createObjectURL(file);
    const video = document.createElement("video");
    video.preload = "metadata";
    video.onloadedmetadata = () => {
      URL.revokeObjectURL(url);
      resolve(Number.isFinite(video.duration) ? video.duration : null);
    };
    video.onerror = () => {
      URL.revokeObjectURL(url);
      resolve(null);
    };
    video.src = url;
  });
}

export async function validateAdStudioPreviewVideoDuration(file: File): Promise<string | null> {
  const seconds = await probeAdStudioVideoDuration(file);
  if (seconds == null) {
    return "Не удалось определить длительность видео";
  }
  if (!isReferenceVideoDurationValid(seconds)) {
    return `Видео должно быть от 2 до ${referenceVideoMaxAllowedSeconds()} сек (сейчас ${seconds.toFixed(1)} сек)`;
  }
  return null;
}
